### ISSUE TITLE
Generic constraints to property of type

### ISSUE THREAD

**Issue by amatiasq** [microsoft/TypeScript#40627]
## Search Terms

generic constraint property type keyof

## Suggestion

To be able to constrain the type of the property of an object using a `keyof` type.

## Use Cases

Code speaks better than me:

```ts
function manageArrayProperty<TObj, TKey extends keyof TObj>(obj: TObj, key: TKey) {
  // obj[key] has to extend array
  obj[key].forEach(...);
}
```

## Examples

This requires a generic declaration that doesn't define a type but only a constraint

```ts
function manageArrayProperty<
  TObj,
  TKey extends keyof TObj,
  TObj[TKey] extends any[]
>(obj: TObj, key: TKey) {
  // this is fine
  obj[key].forEach(...);
}
```

## Previous attempts

I tried to achieve that in different ways, I can create a type constrained to `TObj[TKey]` and `any[]` but that doesn't mean `TObj[TKey]` extends array.

```ts
function manageArrayProperty<
  TObj,
  TKey extends keyof TObj,
  TProp extends TObj[TKey] & any[]
>(obj: TObj, key: TKey) {
  // INVALID: obj[key] can not be an array
  obj[key].forEach(...);
}
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by amatiasq** [comment]
Maybe this should work? I might be missing something

![Screenshot 2020-09-18 at 10 15 49](https://user-images.githubusercontent.com/1533589/93573934-0b690a80-f998-11ea-8e28-361799e7470c.png)

[Playground link](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAdwE4EMAOAeAUIxAFQHkAjAKwBp8iBpAUwE9F6APKesAEwGdEBrJnGBEyVGoQAKqOJhbtOvURQDahBowC6iAGREAkhwC2KzdQKFD9I7gB8ACjgUAXMqoCmr9UwCUiAN40EAg8UIgAbugANiD0XtKyiAC8iE7kKoJaANy4AL64QA)

```ts
function wrap<
  TObj,
  TKey extends keyof TObj,
  TProp extends TObj[TKey] & TItem[],
  TItem
>(obj: TObj, key: TKey) {
  const value: TProp = obj[key];
}
```

**Comment by MartinJohns** [comment]
For all properties:

    function manageArrayProperty<
      TObj extends { [key: string]: string[] },
      TKey extends keyof TObj
    >(obj: TObj, key: TKey) {
      obj[key].forEach(x => {});
    }

For a single property:

    function manageArrayProperty<
      TObj extends { [key in TKey]: string[] },
      TKey extends keyof TObj
    >(obj: TObj, key: TKey) {
      obj[key].forEach(x => {});
    }

Or am I misunderstanding something?

**Comment by amatiasq** [comment]
@MartinJohns no, you understood it perfectly.

I wasn't aware you could do `key in TKey`, thanks!

**Comment by amatiasq** [comment]
Actually @MartinJohns looks like with this approach it's not possible to assign to TObj[TKey] for some reason:

```ts
function manageArrayProperty<
  TObj extends { [key in TKey]: string[] },
  TKey extends keyof TObj
>(obj: TObj, key: TKey) {
  obj[key] = obj[key].map(x => x);
}

const myObject = {
  foo: 1,
  bar: 'test',
  quz: ['a','b','c']
};

manageArrayProperty(myObject, 'quz');
```

[Playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAWwIZlQcwKYEEBO+qAngAr5wAO2+UxAPAFCKIAqA8gEYBWi2AHlGxgAJgGdEAb0QBtANbZiiGElYBpRQF0AXIjFR8KzDM2IAvgBpmbDUoFDREhcTjA2XbowB8ACjg9dDh4LRGdA2wBKKWt-bnktRABeRFj44k0AOjRKH34kr0R+CIBuRjNGRggEfRRiD2xoJOiWYDg4XQBGKxZOVHxdAHIhfQHuxABHEAAvXRkB1FGBzkWIAc0y0sY0DBwCIjIKalpiH2Q6ngaoEIHJqYGSxiA)

Is this a bug maybe?

**Comment by RyanCavanaugh** [comment]
This is a correct error from TypeScript's perspective, because the two utterances of `key` are simply expressions of type `TKey`, and thus not necessarily the same property name, so an invocation where `TKey` was a union of two different property keys could lead to an unsound assignment.

**Comment by amatiasq** [comment]
Thanks for explaining @RyanCavanaugh, I still don't get it though. Particularly when you say "the two utterances of `key`"

On `obj[key] = obj[key].map(x => x)` the two appearances of `key` point to the same reference, they are literally the same value in memory so it can't be different.

If you mean that `key` reference and `key` in `{ [key in TKey]: string[] }`... let's change the name for clarity:

```ts
function manageArrayProperty<
  TObj extends { [p in TKey]: string[] },
  TKey extends keyof TObj
>(obj: TObj, key: TKey) {
  obj[key] = obj[key].map(x => x);
}
```

Then yes, they are different but the `key` variable MUST overlap one of `p` cases. That's how we know `obj[key]` is of type `string[]`.

In other words. `obj[key] = obj[key].map(x => x)` shouldn't be different from `a = a.map(x => x)`.

Even the error message is confusing:

![Screenshot 2020-09-23 at 11 46 15](https://user-images.githubusercontent.com/1533589/93996375-777db100-fd92-11ea-9783-f410018b9e30.png)

```
Type 'string[]' is not assignable to type 'TObj[TKey]'
```

While right above you see 

```
(parameter) obj: TObj extends { [p in TKey]: string[]; }
```

So any `TObj[TKey]` should accept to be assigned a `string[]`.

**Comment by MartinJohns** [comment]
I think this example shows what @RyanCavanaugh means:

    type TestA = { value: 'A'[] }
    type TestB = { value: 'B'[] }

    const testA: TestA = { value: ['A'] }
    const testB: TestB = { value: ['B'] }
    const aOrB: TestA | TestB = Math.random() > 0.5 ? testA : testB;

    manageArrayProperty(aOrB, 'value');

In this case the type written to is **either** `A[]` **or** `B[]`, but the type of the value retrieved is `A[] | B[]`. There is currently no way (AFAIK) to limit a generic type argument to non-union types.

**Comment by amatiasq** [comment]
Let me be clear, even in that case how is that a problem?

According to the function signature `test['a']` and `test['b']` should be of type `string[]`. So inside that function I should be able to even do

```ts
test['b'] = test['a'].map(x => x);
```

Sorry if I'm missing something but I insist, this looks like a bug to me.
